src/auction/src/lib.mo

```diff
@@ -0,0 +1,971 @@
+    #placement : { index : Nat; error : InternalPlaceOrderError };
+  };
+
+  public type CancelOrderError = InternalCancelOrderError or {
+    #UnknownPrincipal;
+  };
+  public type PlaceOrderError = InternalPlaceOrderError or {
+    #UnknownPrincipal;
+  };
+  public type ReplaceOrderError = CancelOrderError or PlaceOrderError;
+
+  public func getTotalPrice(volume : Nat, unitPrice : Float) : Nat = Int.abs(Float.toInt(Float.ceil(unitPrice * Float.fromInt(volume))));
+
+  // internal usage only
+  type OrderCtx = {
+    assetList : (assetInfo : AssetInfo) -> AssocList.AssocList<OrderId, Order>;
+    assetListSet : (assetInfo : AssetInfo, list : AssocList.AssocList<OrderId, Order>) -> ();
+
+    kind : { #ask; #bid };
+    userList : (userInfo : UserInfo) -> AssocList.AssocList<OrderId, Order>;
+    userListSet : (userInfo : UserInfo, list : AssocList.AssocList<OrderId, Order>) -> ();
+
+    oppositeKind : { #ask; #bid };
+    userOppositeList : (userInfo : UserInfo) -> AssocList.AssocList<OrderId, Order>;
+    oppositeOrderConflictCriteria : (orderPrice : Float, oppositeOrderPrice : Float) -> Bool;
+
+    chargeToken : (orderAssetId : AssetId) -> AssetId;
+    chargeAmount : (volume : Nat, price : Float) -> Nat;
+    priorityComparator : (order : Order, newOrder : Order) -> Bool;
+    lowOrderSign : (orderAssetId : AssetId, orderAssetInfo : AssetInfo, volume : Nat, price : Float) -> Bool;
+
+    amountStat : (assetId : AssetId) -> {
+      add : (n : Nat) -> ();
+      sub : (n : Nat) -> ();
+    };
+    volumeStat : (assetId : AssetId) -> {
+      add : (n : Nat) -> ();
+      sub : (n : Nat) -> ();
+    };
+  };
+
+  public class Auction(
+    trustedAssetId : AssetId,
+    settings : {
+      minimumOrder : Nat;
+      minAskVolume : (AssetId, AssetInfo) -> Int;
+      performanceCounter : Nat32 -> Nat64;
+    },
+  ) {
+
+    // a counter of conducted auction sessions
+    public var sessionsCounter = 0;
+    // a counter of ever added order
+    public var ordersCounter = 0;
+    // asset info, index == assetId
+    public var assets : Vec.Vector<AssetInfo> = Vec.new();
+    // user info
+    public let users : RBTree.RBTree<Principal, UserInfo> = RBTree.RBTree<Principal, UserInfo>(Principal.compare);
+    // asset history
+    public var history : List.List<PriceHistoryItem> = null;
+    // stats
+    public var stats : {
+      var usersAmount : Nat;
+      var accountsAmount : Nat;
+      assets : Vec.Vector<{ var bidsAmount : Nat; var totalBidVolume : Nat; var asksAmount : Nat; var totalAskVolume : Nat; var lastProcessingInstructions : Nat }>;
+    } = {
+      var usersAmount = 0;
+      var accountsAmount = 0;
+      assets = Vec.new();
+    };
+
+    public func initUser(p : Principal) : UserInfo {
+      let data : UserInfo = {
+        var currentAsks = null;
+        var currentBids = null;
+        var credits = null;
+        var history = null;
+      };
+      let oldValue = users.replace(p, data);
+      switch (oldValue) {
```

Comment:
Overwritting user is a dangerous action here, because user asks/bids will still be presented in AssetInfo data structure. So it breaks data consistency and will break auction completely if any of their old ask/bid will be fulfilled